doc/src/manual/arrays.md

```diff
@@ -1050,6 +1050,37 @@ is equivalent to `broadcast(f, args...)`, providing a convenient syntax to broad
 ([dot syntax](@ref man-vectorized)). Nested "dot calls" `f.(...)` (including calls to `.+` etcetera)
 [automatically fuse](@ref man-dot-operators) into a single `broadcast` call.
 
+One could intuitively reason about the braodcasting behavior using the following simple "rules";
+they do not strictly match the internal implementation but still provide some insights:
+
+1. If the number of dimensions are different, "add" trailing singleton dimensions to make them the same.
+2. If the axes are different, "extend" the singleton dimensions to make them the same.
```

Comment:
I intentionally use the term "axes" here because Julia allows non-1-based arrays, e.g.,

```julia
julia> using OffsetArrays

julia> Ao = OffsetArray(rand(2), -1)
2-element OffsetArray(::Vector{Float64}, 0:1) with eltype Float64 with indices 0:1:
 0.3365485025378292
 0.5173478379871327

julia> Ao .+ rand(2)
ERROR: DimensionMismatch("arrays could not be broadcast to a common size; got a dimension with lengths 2 and 2")
```

The error message "arrays could not be broadcast to a common size; got a dimension with lengths 2 and 2" is a little bit misleading, though.

Reply:
The first question is whether you want to write the exact rules with all cases, or write a short introduction for people unfamiliar with the concept. I would vote for the latter, and I thought that's what "intuitively reason about" was going for. 

In which case not trying to make the size/axes distinction just reduces the number of things you have to explain. Nothing should be wrong, of course, but non-Array behaviour can be left for later.

(Somewhere there's an issue about this error message, maybe a PR.)